tokio-udev/src/lib.rs

```diff
@@ -137,41 +137,40 @@ impl Stream for MonitorSocket {
 
     fn poll_next(
         self: Pin<&mut Self>,
-        cx: &mut std::task::Context,
+        _: &mut std::task::Context,
     ) -> Poll<Option<Self::Item>> {
-        self.inner.lock().unwrap().poll_receive(cx)
+        self.inner.lock().unwrap().poll_receive()
     }
 }
 
 struct Inner {
-    io: PollEvented<mio_udev::MonitorSocket>,
+    poll: MioPoll,
+    monitor: mio_udev::MonitorSocket,
 }
 
 impl Inner {
-    fn new(monitor: mio_udev::MonitorSocket) -> io::Result<Inner> {
-        Ok(Inner {
-            io: PollEvented::new(monitor)?,
-        })
+    fn new(mut monitor: mio_udev::MonitorSocket) -> io::Result<Inner> {
+        let poll = MioPoll::new()?;
+        poll.registry()
+            .register(&mut monitor, Token(0), Interest::READABLE)?;
+        Ok(Inner { poll, monitor })
     }
 
-    fn poll_receive(
-        &mut self,
-        cx: &mut std::task::Context,
-    ) -> Poll<Option<mio_udev::Event>> {
-        if let Poll::Pending =
-            self.io.poll_read_ready(cx, mio::Ready::readable())
-        {
-            return Poll::Pending;
-        }
-
-        match self.io.get_mut().next() {
-            Some(event) => Poll::Ready(Some(event)),
-            None => {
-                self.io
-                    .clear_read_ready(cx, mio::Ready::readable())
-                    .unwrap();
-                Poll::Pending
+    fn poll_receive(&mut self) -> Poll<Option<mio_udev::Event>> {
+        let mut events = Events::with_capacity(1);
+        self.poll.poll(&mut events, None).unwrap();
```

Comment:
I think here we may wait for [AsyncFd](https://github.com/tokio-rs/tokio/pull/2903) to land on a minor release version of tokio to use that functionality, to avoid allocating memory here and to reduce code to maintain.

If https://github.com/tokio-rs/tokio/pull/2903 doesn't get on a minor release then I'm fine with this being merged as it's the only solution to this problem.

Reply:
@witchof0x20 [AsyncFd](https://docs.rs/tokio/0.3.2/tokio/io/unix/struct.AsyncFd.html) landed on 0.3.2, feel free to update the code with:

```toml
tokio = { version = ">=0.3.2", features = ["net"] }
```

:wink: